### ISSUE TITLE
Introduce flag to allow accessing protected variables in models

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#9059]
@Mathadon has an interesting use case discusse [here](https://github.com/open-ideas/IDEAS/issues/1254#issuecomment-1146089018), where he needs to re-use models that have protected variables to hide the inner details in a different context, where he actually needs to access all their state variables.

Although this is technically forbidden, I see the point of the use case - asking to change the original code and remove the protected keywords doesn't make much sense to me.

@perost, could we add a compiler flag, e.g. --allowProtectedAccess that lifts the constraints that you can't access protected variables from the outside?

**Comment by bilderbuchi** [comment]
Shouldn't such a flag rather be an additional entry in [--allowNonStandardModelica](https://openmodelica.org/doc/OpenModelicaUsersGuide/1.19/omchelptext.html#omcflag-allownonstandardmodelica), since that is, as you say, technically forbidden by the spec?

**Comment by casella** [comment]
> Shouldn't such a flag rather be an additional entry (`nonStdAllowProtectedAccess`?) in [--allowNonStandardModelica](https://openmodelica.org/doc/OpenModelicaUsersGuide/1.19/omchelptext.html#omcflag-allownonstandardmodelica), since the desired behaviour is, as you say, technically forbidden by the spec?

Agreed. BTW, I don't really see a need for using the `nonStd` prefix for the option names, these are arguments to `--allowNonStandardModelica`, so it's already clear it is not standard - I would go for `allowProtectedAccess`.

**Comment by bilderbuchi** [comment]
Agreed, I just kept it consistent with the already present options.

Btw, the proposed solution for the use case cited,

> The reason why I'm protected the sub-models: to avoid generating too large result files, by hiding the most interior components from the end user.

might be an instance of the [XY problem](https://en.wikipedia.org/wiki/XY_problem), where we're trying to improve the workaround (encouraging non-standard Modelica code in the process) instead of solving the root problem. 

If the `HideResult` annotation is not feasible/appropriate in this case (I have not looked at the models in question), maybe the Variable filter output option could be used as-is to only include the variables/states in question. 
Failing that, maybe we can extend the "Variable filter" option with negative filters that _exclude_ all matching variables (those pesky interior components in this case).

**Comment by perost** [comment]
I agree that it would be much better to find an actual solution to the problem, adding a flag is trivial but not very user-friendly.

Looking at the model I can understand why using `HideResult` on every affected variable isn't feasible. I think the most natural solution would be to just put `HideResult=true` on the `vol` component in order to hide everything in it. But the specification isn't really clear on how `HideResult` is supposed to work, and it seems neither Dymola nor OpenModelica supports hiding entire components and their contents. But that might be something to investigate further since we seem to have a clear use case for it.

**Comment by casella** [comment]
I agree that in general we need to find actual solutions to the root problem. However, the problem with all these solutions, besides the non-trivial amount of time to get an agreement with the MA, is that they don't solve the specific user requirement of this case.

As I understand, what @Mathadon is doing is to take a model developed for other purposes and embed it in some kind of super model that has access to its states for filtering options. One shouldn't be forced to spend hours and money modifying the original model in order to achieve that, just because we don't want to make exceptions to the rule.

I am completely opposed to implementing non-standard, possibly undocumented feature of other Modelica tools for which there is no MCP process in place within the MA, because that means shifting to an ideal world where the standard is defined by the MA and all tools implement it, to a law-of-the-jungle model where there the de-facto standard is one implementation that others should follow, which basically means destroying the cooperative model of the MA.

I don't think that is the case here. We have a use case for which making an exception (and accepting all possible consequences) makes sense. Implementing it is trivial. It is not going to become the default behaviour. I really see no problem with that.

**Comment by casella** [comment]
@perost how much time would it take to introduce a flag `--allowNonStandardModelica=allowProtectedAccess`?

**Comment by perost** [comment]
> @perost how much time would it take to introduce a flag `--allowNonStandardModelica=allowProtectedAccess`?

A few minutes once I remember to actually do it. I shortened the flag to just `--allowNonStandardModelica=protectedAccess` though, since "allow" is already part of the flag name.

**Comment by Mathadon** [comment]
@perost I presume that this flag has to be added to the unit tests by @casella and that it's not part of an IDEAS library annotation or something like hat?

**Comment by perost** [comment]
> @perost thanks! I presume that this flag has to be added to the unit tests by @casella and that it's not part of an IDEAS library annotation or something like hat?

I guess there are two options: add the flag to the library testing for IDEAS or add `__OpenModelica_commandLineOptions` annotations to only the affected models. Turning off the check for the whole library is perhaps easiest but has the disadvantage that it allows introducing unintentional errors in the library.

**Comment by casella** [comment]
Thanks @perost!

@Mathadon this is up to you. I can set that flag in the setup of the IDEAS library test on our CI infrastructure (hopefully without breaking it again 😅 ). But then, if you run the model on your PC and that flag is not set, it won't work, which could be annoying.

Alternatively, we can set the flag in a [vendor annotation](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/omedit.html#openmodelica-commandlineoptions-annotation) in every model of you library that needs it, this will look more or less like this:
```modelica
annotation(__OpenModelica_commandLineOptions = "--allowNonStandardModelica=protectedAccess");
```
If there are already annotations there, the `__OpenModelica_commandLineOptions` will be added to them.

I can make a PR with that for you, if you want.

**Comment by Mathadon** [comment]
@casella I prefer to use the vendor annotation. I'll add the myself but if this change is urgent then I don't mind a pull request either since I'm quite busy until my holidays next week!

**Comment by Mathadon** [comment]
This has been added in https://github.com/open-ideas/IDEAS/commit/cbd1414602b65f5f02ab8203417726be51ea478c
